File: tests/fixtures/safe/SecureStretch.java
package demo;

import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.regex.Pattern;
import javax.servlet.http.HttpServletResponse;
import javax.xml.parsers.DocumentBuilderFactory;

public class SecureStretch {

    // Safe: credentials come from the environment
    public String connect() {
        String password = System.getenv("DB_PASSWORD");
        return password;
    }

    // Safe: static path
    public byte[] readConfig() throws Exception {
        return Files.readAllBytes(Paths.get("/etc/app/config.yml"));
    }

    // Safe: user path is normalized and prefix-checked before use
    public byte[] readUpload(String name) throws Exception {
        Path target = Paths.get("/data/uploads/" + name).normalize();
        if (!target.startsWith("/data/uploads")) throw new SecurityException("bad path");
        return Files.readAllBytes(target);
    }

    // Safe: static URL
    public URL healthCheck() throws Exception {
        return new URL("https://status.example.com/health");
    }

    // Safe: strong hash algorithm
    public MessageDigest hashPassword() throws Exception {
        return MessageDigest.getInstance("SHA-256");
    }

    // Safe: non-sensitive log message
    public void logStartup() {
        logger.info("server started");
    }

    // Safe: CSRF protection left enabled, cookies locked down
    public void configure(HttpSecurity http, Cookie cookie) throws Exception {
        http.csrf();
        cookie.setSecure(true);
        cookie.setHttpOnly(true);
    }

    // Safe: static response body, not attacker-controlled input
    public void staticResponse(HttpServletResponse response) throws Exception {
        response.getWriter().write("<html>OK</html>");
    }

    // Safe: cryptographic RNG for a reset token
    public String generatePasswordResetToken() {
        SecureRandom random = new SecureRandom();
        return String.valueOf(random.nextLong());
    }

    // Safe: no nested/overlapping quantifiers
    public Pattern emailPattern() {
        return Pattern.compile("^[a-zA-Z0-9]+@");
    }

    // Safe: redirect target is a fixed, known path
    public void goToLogin(HttpServletResponse response) throws Exception {
        response.sendRedirect("/login");
    }

    // Safe: DOCTYPE declarations disallowed, external DTDs disabled
    public void parseXml(DocumentBuilderFactory dbf) throws Exception {
        dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
        dbf.setFeature("http://apache.org/xml/features/nonvalidating/load-external-dtd", false);
    }
}

File: tests/fixtures/safe/secure-stretch.go
package main

import (
	"crypto/rand"
	"crypto/sha256"
	"crypto/tls"
	"log"
	"net/http"
	"os"
	"regexp"
)

// Safe: credentials come from the environment
func connect() string {
	password := os.Getenv("DB_PASSWORD")
	return password
}

// Safe: static path
func readConfig() ([]byte, error) {
	return os.ReadFile("/etc/app/config.yml")
}

// Safe: static URL
func healthCheck() (*http.Response, error) {
	return http.Get("https://status.example.com/health")
}

// Safe: strong hash algorithm
func hashPassword(password string) [32]byte {
	return sha256.Sum256([]byte(password))
}

// Safe: non-sensitive log message
func logStartup() {
	log.Println("server started")
}

// Safe: TLS verification enabled
func secureClient() *tls.Config {
	return &tls.Config{InsecureSkipVerify: false}
}

// Safe: cryptographic RNG for a session token
func generateSessionToken() ([]byte, error) {
	b := make([]byte, 32)
	_, err := rand.Read(b)
	return b, err
}

// Safe: no nested/overlapping quantifiers
var emailPattern = regexp.MustCompile("^[a-zA-Z0-9]+@")

// Safe: redirect target is a fixed, known path
func goToLogin(w http.ResponseWriter, r *http.Request) {
	http.Redirect(w, r, "/login", http.StatusFound)
}
